src/uterm_input_uxkb.c

```diff
@@ -380,19 +437,40 @@ int uxkb_dev_process(struct uterm_input_dev *dev,
 		     uint16_t key_state, uint16_t code)
 {
 	struct xkb_state *state;
+	struct xkb_compose_state *compose_state;
 	xkb_keycode_t keycode;
 	const xkb_keysym_t *keysyms;
+	xkb_keysym_t one_sym;
 	int num_keysyms, ret;
+	enum xkb_compose_status compose_status;
 	enum xkb_state_component changed;
 
 	if (key_state == KEY_REPEATED)
 		return -ENOKEY;
 
 	state = dev->state;
+	compose_state = dev->compose_state;
 	keycode = code + EVDEV_KEYCODE_OFFSET;
 
 	num_keysyms = xkb_state_key_get_syms(state, keycode, &keysyms);
 
+	one_sym = XKB_KEY_NoSymbol;
+	if (num_keysyms == 1) {
+		/* See: https://bugs.freedesktop.org/show_bug.cgi?id=67167 */
+		one_sym = xkb_state_key_get_one_sym(state, keycode);
+		keysyms = &one_sym;
+	}
+
+	compose_status = XKB_COMPOSE_NOTHING;
+	if (compose_state && key_state == KEY_PRESSED) {
+		/* XKB_KEY_NoSymbol cancels the current compose sequence. */
+		xkb_compose_state_feed(compose_state, one_sym);
+
+		compose_status = xkb_compose_state_get_status(compose_state);
+		if (compose_status == XKB_COMPOSE_COMPOSED)
+			one_sym = xkb_compose_state_get_one_sym(compose_state);
```

Comment:
```suggestion
		if (compose_status == XKB_COMPOSE_COMPOSED) {
			one_sym = xkb_compose_state_get_one_sym(compose_state);
			if (one_sym == XKB_KEY_NoSymbol) {
				/*
				 * It is possible that the sequence only specifies an utf8 string and not a keysym.
				 * Treat this is cancelled, as the rest of the system can not handle it.
				 */
				compose_status = XKB_COMPOSE_CANCELLED;
			}
		}
```

It's possible that `xkb_compose_state_get_one_sym` returns `XKB_KEY_NoSymbol` in case the sequence doesn't define one. Let's be explicit about we are not going to handle the case.

Reply:
I had not thought about that possibility. I've used your code for this.